include/xsparse/levels/hashed.hpp

```diff
@@ -0,0 +1,167 @@
+
+            public:
+                class iterator;
+                using value_type =
+                    typename std::pair<typename BaseTraits::IK, typename BaseTraits::PK>;
+                using difference_type = typename std::make_signed_t<typename BaseTraits::PK>;
+                using pointer =
+                    typename std::pair<typename BaseTraits::IK, typename BaseTraits::PK>*;
+                using reference = std::pair<typename BaseTraits::IK, typename BaseTraits::PK>;
+                using iterator_type = iterator;
+
+                class iterator : xtl::xbidirectional_iterator_base2<iteration_helper>
+                {
+                private:
+                    using wrapped_iterator_type =
+                        typename std::unordered_map<typename BaseTraits::IK,
+                                                    typename BaseTraits::PK>::const_iterator;
+                    wrapped_iterator_type wrapped_it;
+
+                public:
+                    explicit inline iterator(wrapped_iterator_type wrapped) noexcept
+                        : wrapped_it(wrapped)
+                    {
+                    }
+
+                    inline std::tuple<typename BaseTraits::IK, typename BaseTraits::PK> operator*()
+                        const noexcept
+                    {
+                        return { wrapped_it->first, wrapped_it->second };
+                    }
+
+                    inline bool operator==(const iterator& other) const noexcept
+                    {
+                        return wrapped_it == other.wrapped_it;
+                    }
+
+                    inline iterator& operator++() noexcept
+                    {
+                        ++wrapped_it;
+                        return *this;
+                    }
+
+                    inline iterator& operator--() noexcept
+                    {
+                        --wrapped_it;
+                        return *this;
+                    }
+                };
+
+                explicit inline iteration_helper(typename BaseTraits::Level const& level,
+                                                 typename BaseTraits::I const i,
+                                                 typename BaseTraits::PKM1 const pkm1) noexcept
+                    : m_level(level)
+                    , m_pkm1(pkm1)
+                    , m_i(i)
+                {
+                }
+
+                inline iterator begin() const noexcept
+                {
+                    return iterator{ m_level.m_crd[m_pkm1].begin() };
+                }
+
+                inline iterator end() const noexcept
+                {
+                    return iterator{ m_level.m_crd[m_pkm1].end() };
+                }
+            };
+
+            iteration_helper iter_helper(typename BaseTraits::I i, typename BaseTraits::PKM1 pkm1)
+            {
+                return iteration_helper{ *static_cast<typename BaseTraits::Level*>(this), i, pkm1 };
```

Comment:
You can directly pass the relevant map here.

Reply:
Is it okay to remove `BaseTraits::I i` as a parameter as it is not used anywhere?

Reply:
Yes, remove everything but the map.